apps/aecore/lib/aecore/channel/channel_state_peer.ex

```diff
@@ -46,118 +52,362 @@ defmodule Aecore.Channel.ChannelStatePeer do
+         }"}
+
+      # check sequence
+      tx_sequence <= cur_sequence ->
+        {:error, "#{__MODULE__}: Invalid sequence in tx"}
+
+      true ->
+        :ok
+    end
+  end
+
+  @spec update_offchain_chainstate(ChannelTransaction.signed_tx(), ChannelStatePeer.t()) ::
+          {:ok, Chainstate.t()} | error()
+  defp update_offchain_chainstate(tx, %ChannelStatePeer{
+         channel_reserve: channel_reserve,
+         offchain_chainstate: offchain_chainstate
+       }) do
+    unsigned_tx = ChannelTransaction.unsigned_payload(tx)
+    state_hash = unsigned_tx.state_hash
+
+    with {:ok, updated_offchain_chainstate} <-
+           ChannelOffChainUpdate.apply_updates(
+             offchain_chainstate,
+             ChannelTransaction.offchain_updates(tx),
+             channel_reserve
+           ),
+         ^state_hash <- Chainstate.calculate_root_hash(updated_offchain_chainstate) do
+      {:ok, updated_offchain_chainstate}
+    else
+      {:error, _} = err ->
+        err
+
+      hash when is_binary(hash) ->
+        {:error,
+         "#{__MODULE__}: Wrong state hash in tx, expected #{inspect(hash)}, got: #{
+           inspect(state_hash)
+         }"}
+    end
+  end
 
   @doc """
-  Creates a channel from a signed channel state.
+  Creates a channel from a list of mutually signed tx. The first tx in the list must be ChannelCreateTX. All the tx and updates are verified for corectness along the way.
   """
-  @spec from_state(
-          ChannelStateOffChain.t(),
-          Keys.pubkey(),
-          Keys.pubkey(),
-          non_neg_integer(),
+  @spec from_signed_tx_list(
+          list(ChannelTransaction.signed_tx()),
           Channel.role()
-        ) :: ChannelStatePeer.t()
-  def from_state(state, initiator_pubkey, responder_pubkey, channel_reserve, role) do
-    %ChannelStatePeer{
+        ) :: {:ok, ChannelStatePeer.t()} | error()
+  def from_signed_tx_list(offchain_tx_list, role) do
+    offchain_tx_list_from_oldest = Enum.reverse(offchain_tx_list)
+
+    [create_tx | _] = offchain_tx_list_from_oldest
+
+    %SignedTx{
+      data:
+        %DataTx{
+          type: ChannelCreateTx,
+          payload: %ChannelCreateTx{
+            channel_reserve: channel_reserve
+          },
+          senders: [
+            %Identifier{value: initiator_pubkey},
+            %Identifier{value: responder_pubkey}
+          ]
+        } = data_tx
+    } = create_tx
+
+    channel_id = ChannelStateOnChain.id(data_tx)
+
+    initial_state = %ChannelStatePeer{
       fsm_state: :open,
       initiator_pubkey: initiator_pubkey,
       responder_pubkey: responder_pubkey,
       role: role,
-      highest_state: state,
-      highest_signed_state: state,
+      channel_id: channel_id,
       channel_reserve: channel_reserve
     }
+
+    Enum.reduce_while(offchain_tx_list_from_oldest, {:ok, initial_state}, fn tx, {:ok, state} ->
+      case process_fully_signed_tx(tx, state) do
+        {:ok, _} = new_acc ->
+          {:cont, new_acc}
```

Comment:
what is `:cont` maybe use a more semantic name

Reply:
This is for `reduce_while`. `reduce_while` compared to `reduce` stops also when there are no more items to process or supplied function returns `{:halt, value}`

Reply:
ah okay, I didn't know